lib/spack/spack/cmd/config.py

```diff
@@ -431,6 +443,79 @@ def config_revert(args):
         tty.msg(msg.format(cfg_file))
 
 
+def config_prefer_upstream(args):
+    """Generate a packages config based on the configuration of all upstream
+    installs."""
+
+    scope = args.scope
+    if scope is None:
+        scope = spack.config.default_modify_scope('packages')
+
+    all_specs = set(spack.store.db.query(installed=True))
+    local_specs = set(spack.store.db.query_local(installed=True))
+    pref_specs = local_specs if args.local else all_specs - local_specs
+
+    conflicting_variants = set()
+
+    pkgs = {}
+    for spec in pref_specs:
+        # Collect all the upstream compilers and versions for this package.
+        pkg = pkgs.get(spec.name, {
+            'version': [],
+            'compiler': [],
+        })
+        pkgs[spec.name] = pkg
+
+        # We have no existing variant if this is our first added version.
+        existing_variants = pkg.get('variants',
+                                    None if not pkg['version'] else '')
+
+        version = spec.version.string
+        if version not in pkg['version']:
+            pkg['version'].append(version)
+
+        compiler = str(spec.compiler)
+        if compiler not in pkg['compiler']:
+            pkg['compiler'].append(compiler)
+
+        # Get and list all the variants that differ from the default.
+        variants = []
+        for var_name, variant in spec.variants.items():
+            if (var_name in ['patches']
+                    or var_name not in spec.package.variants):
+                continue
+
+            if variant.value != spec.package.variants[var_name].default:
+                variants.append(str(variant))
+        variants.sort()
+        variants = ' '.join(variants)
+
+        if spec.name not in conflicting_variants:
+            # Only specify the variants if there's a single variant
+            # set across all versions/compilers.
+            if existing_variants is not None and existing_variants != variants:
+                conflicting_variants.add(spec.name)
+                del pkg['variants']
```

Comment:
```suggestion
                pkg.pop('variants')
```
If the first variants setting was empty, we don't have anything to delete from the dictionary when we reach a conflict with the second one.